tree-sitter-highlight.el

```diff
@@ -0,0 +1,125 @@
+                                                     ))
+                       (buffer-string))))
+    ;; TODO: figure out why ts-make-query doesn't work
+    (ts--make-query tree-sitter-language highlights)))
+
+(defun tree-sitter-highlight--apply-match (match)
+  "Apply the given MATCH to the buffer."
+  (let* ((name (aref match 0))
+         (node (aref match 1))
+         (face (or (gethash name tree-sitter-highlight--current-face-hash)
+                   (gethash (car (split-string name "\\.")) tree-sitter-highlight--current-face-hash)))
+         (start (ts-node-start-position node))
+         (end   (ts-node-end-position node)))
+    ;; (set-text-properties start end nil)
+    (add-face-text-property start end face)))
+
+(defun tree-sitter-highlight--handle-change (change)
+  "Called for each CHANGE after an edit."
+  (let* ((start-byte (aref change 0))
+         (end-byte   (aref change 1))
+         (node       (ts-get-descendant-for-byte-range (ts-root-node tree-sitter-tree) start-byte end-byte))
+         ;; TODO: why isn't #'ts-node-text chosen as the default, the comment says it should be?
+         (matches    (ts-query-captures tree-sitter-highlight--queries node tree-sitter-highlight--query-cursor nil #'ts-node-text)))
+    (set-text-properties (ts-node-start-position node) (ts-node-end-position node) nil)
+    (seq-do #'tree-sitter-highlight--apply-match matches)))
+
+(defun tree-sitter-highlight--after-change (old-tree)
+  "Called with the OLD-TREE after the buffer has changed."
+  (let* ((changes (ts-changed-ranges tree-sitter-tree old-tree)))
+    (seq-do #'tree-sitter-highlight--handle-change changes)))
+
+(defun tree-sitter-highlight--highlight ()
+  "Highlights the whole buffer."
+  (let* ((node (ts-root-node tree-sitter-tree))
+         (matches (ts-query-captures tree-sitter-highlight--queries node tree-sitter-highlight--query-cursor nil #'ts-node-text)))
+    (seq-do #'tree-sitter-highlight--apply-match matches)))
+
+(defun tree-sitter-highlight--enable ()
+  "Enable `tree-sitter-highlight' in this buffer."
+  (setq tree-sitter-highlight--current-face-hash (alist-get major-mode tree-sitter-highlight-face-hash))
+  ;; TODO: error handling for the above line.
+  (setq tree-sitter-highlight--queries (tree-sitter-highlight--make-queries))
+  (setq tree-sitter-highlight--query-cursor (ts-make-query-cursor))
+  (add-hook 'tree-sitter-after-change-functions #'tree-sitter-highlight--after-change 'append 'local)
+  (tree-sitter-highlight--highlight)
+  )
+
+(defun tree-sitter-highlight--disable ()
+  "Disable `tree-sitter-highlight' in this buffer."
+  (remove-hook 'tree-sitter-after-change-functions #'tree-sitter-highlight--after-change 'local)
+  (set-text-properties (point-min) (point-max) nil)
+  )
+
+(define-minor-mode tree-sitter-highlight-mode
+  "Syntax highlighting with tree sitter."
+  :init-value nil
+  :lighter "tree-sitter-highlight"
+  ;; TODO: idk if this file should do this.
```

Comment:
Dependencies between minor modes are tricky: https://emacs.stackexchange.com/questions/36225/global-minor-mode-that-requires-another-during-its-lifetime.

We can either use a single minor mode, or make `tree-sitter-mode` aware of its possible dependent `tree-sitter-` minor modes, so that turning them on/off work properly.